### PR TITLE
DM-25475: Implement M2 LUT behavior in m2 simulator

### PR DESCRIPTION
**Pull request by tribeiro**
Make simulated pga listen for telescope position to emulate inclinometer.

### CODE REVIEW

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 1]
```diff
@@ -118,8 +151,18 @@ void SimulatedFPGA::pullTelemetry() {
...
+    // This comment is made in src/LSST/M1M3/SS/Inclinometer/Inclinometer.cpp
+    // 0 = Zenith, we want 0 to = Horizon
+    // Also we want below horizon to be negative
+    // Which I take to mean that the inclinometer is measuring Zenith angle,
+    // so I will convert elevation to zenith angle here.
+
+    this->elevationReadWriteLock.lock();
     this->supportFPGAData.InclinometerAngleRaw =
-            (int32_t)(tmaElevation.Elevation_Angle_Actual * 1000.0) + (this->getRnd() * 5.0);
+            (int32_t)((this->mountElevation - 90.0) * 1000.0) + (this->getRnd() * 5.0);
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:163]: This should be (90 - this->mountElevation) * 1000. By trivial check, elevation angle 45 deg is zenith angle 45 deg. Your code will show -45 (which technicaly is the same, but I do prefer positive values for zenith angle measurements)

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 2]
```diff
@@ -118,8 +151,18 @@ void SimulatedFPGA::pullTelemetry() {
     this->supportFPGAData.InclinometerErrorTimestamp = 0;
     this->supportFPGAData.InclinometerErrorCode = 0;
     this->supportFPGAData.InclinometerSampleTimestamp = timestamp;
+
+    // This comment is made in src/LSST/M1M3/SS/Inclinometer/Inclinometer.cpp
+    // 0 = Zenith, we want 0 to = Horizon
+    // Also we want below horizon to be negative
+    // Which I take to mean that the inclinometer is measuring Zenith angle,
+    // so I will convert elevation to zenith angle here.
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:159]: Don't add more than is necessary to comment. Just state:

```
Inclinometer raw value is measured as zenith angle. Converts elevation to zenith angle.
```

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 3]
```diff
@@ -118,8 +151,18 @@ void SimulatedFPGA::pullTelemetry() {
...
+
+    this->elevationReadWriteLock.lock();
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:161]: use std::lock_guard, as such:

```cpp
{
   std::lock_guard<std::mutex> lock_g(this->elevationReadWriteLock);
   code..
}
```

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 4]
```diff
@@ -118,8 +151,18 @@ void SimulatedFPGA::pullTelemetry() {
...
     this->supportFPGAData.InclinometerAngleRaw =
-            (int32_t)(tmaElevation.Elevation_Angle_Actual * 1000.0) + (this->getRnd() * 5.0);
+            (int32_t)((this->mountElevation - 90.0) * 1000.0) + (this->getRnd() * 5.0);
+    this->elevationReadWriteLock.unlock();
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:164]: and of course remove this when using lock_guard

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 5]
```diff
@@ -49,6 +52,36 @@ SimulatedFPGA::SimulatedFPGA() {
         this->rnd[i] = float((rand() % 2000) - 1000) / 1000.0;
     }
     this->rndIndex = 0;
+    this->mgr_MTMount = SAL_MTMount();
+    this->mgr_MTMount.salTelemetrySub(const_cast<char*>("MTMount_Elevation"));
+
+    this->elevationReadWriteLock.unlock();
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:58]: mutex come unlocked (by constructor). Please remove this extra line.

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 6]
```diff
@@ -33,6 +33,9 @@
 #include <unistd.h>
 #include <cstdlib>
 #include <spdlog/spdlog.h>
+#include <thread>
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:36]: you included thread in .h. No need to include it here again

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 7]
```diff
@@ -49,6 +52,36 @@ SimulatedFPGA::SimulatedFPGA() {
...
+    while (!this->exitThread) {
+        ReturnCode_t status = mgr_MTMount.getSample_Elevation(&mountElevationInstance);
+
+        if (status == 0) {
+            spdlog::debug("Got valid elevation sample...");
+            this->elevationReadWriteLock.lock();
+            this->mountElevation = mountElevationInstance.Elevation_Angle_Actual;
+            this->elevationReadWriteLock.unlock();
+        }
+        usleep(20000);
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:83]: usleep doesn't work with std::thread run (or some advanced thread options I forgot..but I remember I spend significant time debugging that with LBTO multithreads). Please replace with std::this_thread::sleep_for(20ms) - https://en.cppreference.com/w/cpp/thread/.

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 8]
```diff
@@ -49,6 +52,36 @@ SimulatedFPGA::SimulatedFPGA() {
...
+    MTMount_ElevationC mountElevationInstance;
+
+    spdlog::debug("Start monitoring mount elevation...");
+
+    while (!this->exitThread) {
+        ReturnCode_t status = mgr_MTMount.getSample_Elevation(&mountElevationInstance);
+
+        if (status == 0) {
+            spdlog::debug("Got valid elevation sample...");
+            this->elevationReadWriteLock.lock();
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:79]: Please replace lock/unlock with C++11 lock_guard

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.h [hunk 9]
```diff
@@ -106,7 +113,12 @@ class SimulatedFPGA : public IFPGA {
     void writeModbusCRC(std::queue<uint16_t>* response);
     uint16_t readModbus(uint16_t data);
 
+    void monitorElevation(void);
+
+    int exitThread = false;
     float rnd[RND_CNT];
+    float mountElevation = 90.;
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.h:120]: private values and methods shall be prefixed with _ according to LSST DM C++ developer guide. Actually there is a patch (WIP) to fix this in full M1M3 SS code. When adding new private values, please prefix with _ and don't use this-> in the code to access them.

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp (hunk 1 above)
Comment (tribeiro) [reply src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:163]: yes, but check the math done here: https://github.com/lsst-ts/ts_m1m3support/blob/c8c90ba18b2351963a69ff810f2834dbfd6cac19/src/LSST/M1M3/SS/Inclinometer/Inclinometer.cpp#L93

I think it is wrong there instead. Should I correct this? I am afraid to change it there as this was tested with the actual hardware and I am not sure if it is used or not. If I correct the math here I have to correct it there as well...

Comment (tribeiro) [reply src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:163]: Just to be clear. If I do 

```
InclinometerAngleRaw = (90 - mountElevation) * 1000.
```

which is the correct, then,

```
angle = 90.0 - InclinometerAngleRaw / 1000.0
```

which is not what `inclinometer.cpp` is computing... So either I correct both places or leave it the way it is in both places.. But `inclinometer.cpp` was tested with real hardware... so I really don't know if this is how it is supposed to be or a previous mistake.

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 10]
```diff
@@ -49,6 +51,38 @@ SimulatedFPGA::SimulatedFPGA() {
...
+
+        if (status == 0) {
+            spdlog::debug("Got valid elevation sample...");
+
+            {
+                std::lock_guard<std::mutex> lock_g(_elevationReadWriteLock);
+                _mountElevation = mountElevationInstance.Elevation_Angle_Actual;
+            }
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(20000));
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:84]: It was originally usleep(20000). Should now be std::chrono::milliseconds(20) - usleep sleeps for microseconds, which is ms * 1000

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp [hunk 11]
```diff
@@ -118,8 +152,15 @@ void SimulatedFPGA::pullTelemetry() {
     this->supportFPGAData.InclinometerErrorTimestamp = 0;
     this->supportFPGAData.InclinometerErrorCode = 0;
     this->supportFPGAData.InclinometerSampleTimestamp = timestamp;
-    this->supportFPGAData.InclinometerAngleRaw =
-            (int32_t)(tmaElevation.Elevation_Angle_Actual * 1000.0) + (this->getRnd() * 5.0);
+
+    // Inclinometer raw value is measured as zenith angle. Converts elevation to zenith angle.
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.cpp:156]: Please  change that to:

// Inclinometer raw value is measured as (negative) zenith angle (0 = zenith, -90 = horizon). Converts elevation to zenith angle and adds random 5 deg noise

File: src/LSST/M1M3/SS/FPGA/SimulatedFPGA.h [hunk 12]
```diff
@@ -106,7 +113,12 @@ class SimulatedFPGA : public IFPGA {
     void writeModbusCRC(std::queue<uint16_t>* response);
     uint16_t readModbus(uint16_t data);
 
+    void _monitorElevation(void);
+
+    int _exitThread = false;
```
Comment (pkubanek) [inline src/LSST/M1M3/SS/FPGA/SimulatedFPGA.h:118]: I actually missed that - changed to bool (from int) in next commit

**Pull request merged**